### PR TITLE
Map costs

### PR DESCRIPTION
**Pull request by NautiluX**
Query hive cluster to get mapping of AWS Accounts to Account and AccountClaim CRs.

jira: [OSD-10006](https://issues.redhat.com/browse/OSD-10006)

Support querying root OUs

jira: [OSD-9906](https://issues.redhat.com/browse/OSD-9906)

### CODE REVIEW

File: cmd/cost/list.go [hunk 1]
```diff
@@ -270,6 +309,10 @@ func (o OUCost) printCostPerAccount(awsClient awsprovider.Client) {
 			Unit:      accountCost.Unit,
 		}
 		if o.options.csv {
+			if o.options.claims {
+				fmt.Printf("%s,%s,%s,%s,%s,%s\n", accountCost.accountClaimNamespace, accountCost.accountClaimName, *o.OU.Id, accountCost.AccountID, accountCost.Cost.StringFixed(2), accountCost.Unit)
```
Comment (dkeohane) [inline cmd/cost/list.go:313]: Could encompass this into a ToString method in accountCost?

**Review by dkeohane — COMMENTED** [review]
Changes look good as is, just minor comments / ideas :t-rex:

File: cmd/cost/list.go [hunk 2]
```diff
@@ -259,8 +275,31 @@ func (o OUCost) getSum() (sum decimal.Decimal, unit string, err error) {
...
 
+func (o *OUCost) getAccountClaims(kubeCli client.Client) {
+	for i, cost := range o.Costs {
+		accountclaim, err := accountget.GetAccountClaimFromAccountID(context.TODO(), o.options.kubeCli, cost.AccountID, common.AWSAccountNamespace)
+		if err == nil {
+			o.Costs[i].accountClaimName = accountclaim.Name
+			o.Costs[i].accountClaimNamespace = accountclaim.Namespace
+			continue
+		}
+		if strings.Contains(fmt.Sprint(err), "AccountClaim matched for Account CR") {
```
Comment (macgregor) [inline cmd/cost/list.go:286]: why not just have the underlying error string be "AccountClaim CR not found" or "Account CR not found" (the string you want displayed) so all you need is line 294?

Also side note, is this a common Go pattern for handling different error types? Im coming from a Java background where we would would generally define a custom exception (e.g. `class AccountClaimCRNotFound extends Exception` and `class AccountCRNotFound extends Exception`) so you can get strong type checks.

Comment (NautiluX) [reply cmd/cost/list.go:286]: A error class hierarchy is not very common in go, and often you'll find yourself figure out the error type by string comparison as a quick solution. it feels hacky, but implementing better tooling often is a lot of effort. The k8s packages have a methodology for error checking where you can can call e.g. `isNotFound` to figure out if an error returned by the API servier is a 404.

File: cmd/account/get/account-claim.go [hunk 3]
```diff
@@ -136,3 +109,38 @@ func (o *getAccountClaimOptions) run() error {
...
+		return
+	}
+
+	for _, a := range accountClaims.Items {
+		if a.Spec.AccountLink == accountName {
+			accountClaim = a
+			return
+		}
+	}
+	err = fmt.Errorf("AccountClaim matched for Account CR %s not found\n", accountName)
```
[the remark this answers is not in the document]
Comment (NautiluX) [reply cmd/account/get/account-claim.go:144]: yup I'd say let's not do this now...

File: cmd/account/get/account-claim.go [hunk 4]
```diff
@@ -136,3 +109,38 @@ func (o *getAccountClaimOptions) run() error {
...
+func GetAccountClaimFromAccountID(ctx context.Context, kubeCli client.Client, accountID string, namespace string) (accountClaim awsv1alpha1.AccountClaim, err error) {
+	accounts := awsv1alpha1.AccountList{}
+	err = kubeCli.List(ctx, &accounts, &client.ListOptions{
+		Namespace: namespace,
+	})
+	if err != nil {
+		return
+	}
+
+	for _, a := range accounts.Items {
```
[the remark this answers is not in the document]
Comment (NautiluX) [reply cmd/account/get/account-claim.go:122]: agree

File: cmd/account/get/account-claim.go [hunk 5]
```diff
@@ -136,3 +109,38 @@ func (o *getAccountClaimOptions) run() error {
...
+		if a.Spec.AwsAccountID == accountID {
```
[the remark this answers is not in the document]
Comment (NautiluX) [reply cmd/account/get/account-claim.go:134]: done

File: cmd/account/get/account-claim.go [hunk 6]
```diff
@@ -134,5 +110,61 @@ func (o *getAccountClaimOptions) run() error {
...
+		Namespace: namespace,
+	})
+	if err != nil {
+		return
+	}
+
+	errors := []error{}
+	for _, a := range accounts.Items {
+		if a.Spec.AwsAccountID == accountID {
+			claim, claimerr := getAccountClaim(ctx, kubeCli, a.Name)
```
Comment (dkeohane) [inline cmd/account/get/account-claim.go:135]: minor: We could save a loop here when getting the account claim by using the name and namespace in the search - in the account CR object we store them in the claimLink and claimLinkNamespace fields

Comment (NautiluX) [reply cmd/account/get/account-claim.go:135]: agree this could be done

File: cmd/cost/list.go (hunk 1 above)
Comment (NautiluX) [reply cmd/cost/list.go]: yes we could do this, I'll see if I get to following up on it.

File: cmd/account/get/account-claim.go [hunk 7]
```diff
@@ -134,5 +110,61 @@ func (o *getAccountClaimOptions) run() error {
 		return err
 	}
 
-	return resourcePrinter.PrintObj(&accountClaim, o.Out)
+	for _, accountClaim := range accountClaims {
+		err = resourcePrinter.PrintObj(&accountClaim, o.Out)
+		if err != nil {
+			return nil
```
[the remark this answers is not in the document]
Comment (NautiluX) [reply cmd/account/get/account-claim.go]: fixed

**Review by dkeohane — APPROVED** [review]
/lgtm

File: cmd/cost/list.go [hunk 8]
```diff
@@ -281,7 +368,11 @@ func (o OUCost) printCostPerAccount(awsClient awsprovider.Client) {
 	}
 	if o.options.csv {
 		if o.options.sum {
-			fmt.Printf("%s,%s,%s,%s\n", *o.OU.Id, "SUM", sum.StringFixed(2), unit)
+			output := fmt.Sprintf("%s,%s,%s,%s", *o.OU.Id, "SUM", sum.StringFixed(2), unit)
+			if o.options.claims {
+				output = ",," + output
```
Comment (dastergon) [inline cmd/cost/list.go:373]: Are `,,` as output intentional?

Comment (dastergon) [reply cmd/cost/list.go:373]: Would it make sense to add the format as a comment?

File: cmd/cost/create.go [hunk 9]
```diff
@@ -60,6 +62,12 @@ func createCostCategory(OUid *string, OU *organizations.OrganizationalUnit, awsC
 		return err
 	}
 
+	accountIdList := []*string{}
+	for i := range accounts {
+		accountIdList = append(accountIdList, &accounts[i].accountId)
+
```
Comment (dastergon) [inline cmd/cost/create.go:68]:
```suggestion
```

File: cmd/cost/list.go (hunk 8 above)
Comment (NautiluX) [reply cmd/cost/list.go:373]: it's CSV generation. Not very beautiful, but works.

**Comment by iamkirkbater** [comment]
Closing in favor of continuing development in #219

**Pull request closed without merging**